lib/routemaster/fetcher.rb

```diff
@@ -28,23 +27,45 @@ def initialize(middlewares: [], listener: nil)
     # string otherwise.
     def get(url, params: {}, headers: {})
       host = URI.parse(url).host
-      connection.get(url, params, headers.merge(auth_header(host)))
+      response_wrapper do
+        connection.get(url, params, headers.merge(auth_header(host)))
+      end
+    end
+
+    def post(url, body: {}, headers: {})
+      host = URI.parse(url).host
+      response_wrapper do
+        connection.post do |req|
+          req.url url
+          req.headers = headers.merge(auth_header(host))
+          req.body = body.to_json
+        end
+      end
+    end
+
+    def discover(url)
+      get(url)
     end
 
     private
 
-    def generate_middlewares(middlewares)
-      default_middleware = [DEFAULT_MIDDLEWARE, listener: @listener]
-      middlewares.unshift(default_middleware)
+    def response_wrapper(&block)
+      response = block.call
+      @response_class ? @response_class.new(response) : response
     end
 
     def connection
       @connection ||= Faraday.new do |f|
         f.request  :retry, max: 2, interval: 100e-3, backoff_factor: 2
         f.response :mashify
         f.response :json, content_type: /\bjson/
+        f.use Routemaster::Middleware::Caching, listener: @listener
```

Comment:
I think the naming here is not ideal. The cache middleware not "Caching" is using Fetcher and fetcher is using a "Caching" middleware != "Cache" to support response caching. Maybe we can rename the caching middleware to ResponseCaching? 

Do i make any sense here? :D